api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirImmunizationServiceImpl.java

```diff
@@ -0,0 +1,139 @@
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import ca.uhn.fhir.rest.api.SortSpec;
+import ca.uhn.fhir.rest.param.ReferenceParam;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang.StringUtils;
+import org.hl7.fhir.r4.model.Immunization;
+import org.openmrs.Concept;
+import org.openmrs.Obs;
+import org.openmrs.Patient;
+import org.openmrs.api.ConceptService;
+import org.openmrs.api.EncounterService;
+import org.openmrs.api.ObsService;
+import org.openmrs.api.PatientService;
+import org.openmrs.module.fhir2.api.FhirImmunizationService;
+import org.openmrs.module.fhir2.api.translators.ImmunizationTranslator;
+import org.openmrs.module.fhir2.api.translators.impl.ImmunizationObsGroupHelper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FhirImmunizationServiceImpl implements FhirImmunizationService {
+	
+	@Autowired
+	private ImmunizationTranslator translator;
+	
+	@Autowired
+	private ObsService obsService;
+	
+	@Autowired
+	private EncounterService encounterService;
+	
+	private PatientService patientService;
+	
+	private ImmunizationObsGroupHelper helper;
+	
+	@Autowired
+	public void setPatientService(PatientService patientService) {
+		this.patientService = patientService;
+	}
+	
+	@Autowired
+	public void setImmunizationHelper(ConceptService conceptService) {
+		this.helper = new ImmunizationObsGroupHelper(conceptService);
+	}
+	
+	@Override
+	public Immunization get(String uuid) {
+		Obs obs = obsService.getObsByUuid(uuid);
+		return translator.toFhirResource(obs);
+	}
+	
+	@Override
+	public Immunization create(Immunization newImmunization) {
+		Obs obs = translator.toOpenmrsType(newImmunization);
+		if (obs.getEncounter().getId() == null) {
+			encounterService.saveEncounter(obs.getEncounter());
+		}
+		obs = obsService.saveObs(obs, "Created when translating a FHIR Immunization resource.");
+		newImmunization = translator.toFhirResource(obs);
+		return newImmunization;
+	}
+	
+	@Override
+	public Immunization update(String uuid, Immunization updatedImmunization) {
+		Obs existingObs = obsService.getObsByUuid(uuid);
+		Obs obs = translator.toOpenmrsType(existingObs, updatedImmunization);
+		obs = obsService.saveObs(obs, "Updated when translating a FHIR Immunization resource.");
+		return translator.toFhirResource(obs);
+	}
+	
+	@Override
+	public Immunization delete(String uuid) {
+		// TODO Simply void the underlying obs
+		throw new UnsupportedOperationException("Deleting a FHIR Immunization resource is currently not supported.");
+	}
+	
+	@Override
+	public Concept getOpenmrsImmunizationConcept() {
+		return helper.concept(immunizationGroupingConcept);
+	}
+	
+	public Patient getPatient(ReferenceParam patientParam) {
```

Comment:
I get the motivation here with the `IllegalArgumentException`s, but they're going to provide a sub-optimal experience for the FHIR client trying to consume these. If we are going to throw exceptions at this point, it's probably best to use HAPI's `UnprocessableEntityException` for this, which results in a FHIR-readable error with an appropriate structure including the message.

Reply:
This method will be gone :-)
I may have used `IllegalArgumentExceptions` in other places though, we'll see.